hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -673,95 +755,85 @@ private class Emit(
+            Code(xElt := elt,
+              tmpAcc := codeB.map(v => accType.copyFromPValue(mb, region, PCode(body.pType, codeB.v))),
+              xAcc := tmpAcc)
           }
 
           val codeZ = emit(zero).map(accType.copyFromPValue(mb, region, _))
-          def retTT(acc: TypedTriplet[accType.type]): Code[Ctrl] =
-            ret(COption.fromEmitTriplet(acc.untyped))
+          def retTT(): Code[Ctrl] =
+            ret(COption.fromEmitCode(xAcc.get))
 
-          stream.stream.map(TypedTriplet(eltType, _))
-                .fold(mb)(TypedTriplet(accType, codeZ), foldBody, retTT)
+          stream.stream
+                .fold(mb, xAcc := codeZ, foldBody, retTT())
         }
 
-        COption.toEmitTriplet(resOpt, accType, mb)
+        COption.toEmitCode(resOpt, accType, mb)
 
       case x@StreamFold2(a, acc, valueName, seq, res) =>
         val eltType = coerce[PStream](a.pType).elementType
-        val eltPack = TypedTriplet.pack(eltType)
-
-        val accPacks = x.accPTypes.map(TypedTriplet.pack(_))
-        val accsPack = ParameterPack.array(accPacks)
-        implicit val ap = accsPack.asInstanceOf[ParameterPack[IndexedSeq[TypedTriplet[_]]]]
 
-        val xElt = eltPack.newFields(mb.fb, valueName)
+        val xElt = mb.newEmitField(valueName, eltType)
         val names = acc.map(_._1)
-        val accVars = accsPack.newFields(mb.fb, names)
+        val accVars = (names, x.accPTypes).zipped.map(mb.newEmitField)
+        val tmpAccVars = (names, x.accPTypes).zipped.map(mb.newEmitField)
 
-        val resEnv = Emit.bindEnv(env, names.zip(accVars.pss.asInstanceOf[IndexedSeq[ParameterStoreTriplet[_]]]): _*)
-        val seqEnv = Emit.bindEnv(resEnv, valueName -> xElt)
+        val resEnv = env.bind(names.zip(accVars): _*)
+        val seqEnv = resEnv.bind(valueName, xElt)
 
-        val zero = acc.zip(x.accPTypes).map { case ((_, value), unifyT) =>
-          TypedTriplet(unifyT, emit(value).map(unifyT.copyFromPValue(mb, region, _)))
-        }
         val codeR = emit(res, env = resEnv)
-        val typedCodeSeq = seq.map(ir => TypedTriplet(ir.pType, emit(ir, env = seqEnv)))
+        val typedCodeSeq = seq.map(ir => emit(ir, env = seqEnv))
 
         val streamOpt = emitStream(a)
-
-        val resOpt = streamOpt.flatMapCPS[Code[_]] { (stream, _ctx, ret) =>
+         val resOpt = streamOpt.flatMapCPS[PCode] { (stream, _ctx, ret) =>
           implicit val c = _ctx
 
-          def foldBody(elt: TypedTriplet[eltType.type], accs: IndexedSeq[TypedTriplet[_]], k: IndexedSeq[TypedTriplet[_]] => Code[Ctrl]): Code[Ctrl] =
+          def foldBody(elt: EmitCode): Code[Unit] =
             Code(xElt := elt,
-              accVars := accs.zip(x.accPTypes).map { case (acc, unifyT) =>
-                TypedTriplet(unifyT, acc.untyped.map(unifyT.copyFromPValue(mb, region, _)))
-              },
-              k(typedCodeSeq))
+              tmpAccVars := typedCodeSeq.map(_.copyToRegion(mb, region)),
```

Comment:
I think this copyToRegion stuff is wrong -- it doesn't coerce to the accumulator types. Don't we need to do a `copyFromPValue` instead?